### PR TITLE
refactor: RSI: abstract out average calculation

### PR DESCRIPTION
**Pull request by tlsim**
* Abstract out average calculation
* Calculate changes using [prevClose, close] pairs

### CODE REVIEW

File: src/relativeStrengthIndex.js [hunk 1]
```diff
@@ -8,41 +8,48 @@ export default function() {
...
-        const downChanges = [];
-        const upChanges = [];
-
-        closes.forEach(close => {
-            const downChange = prevClose < close ? 0 : prevClose - close;
-            const upChange = prevClose > close ? 0 : close - prevClose;
-
-            downChanges.push(downChange);
-            upChanges.push(upChange);
+    const downChange = ([prevClose, close]) => prevClose < close ? 0 : prevClose - close;
```
Comment (tlsim) [inline src/relativeStrengthIndex.js]: Babel inlines `babelHelpers.slicedToArray` in the output to support this. Should be ok?

File: src/relativeStrengthIndex.js [hunk 2]
```diff
@@ -8,41 +8,48 @@ export default function() {
...
+    const upChange = ([prevClose, close]) => prevClose > close ? 0 : close - prevClose;
+
+    const makeAverager = prevAverage => changes => {
+        const average = prevAverage ? wildersSmoothing(changes, prevAverage)
+            : changes.reduce(sum) / changes.length;
```
Comment (chrisprice) [inline src/relativeStrengthIndex.js]: https://github.com/d3/d3/wiki/Arrays#d3_mean?

**Comment by tlsim** [comment]
@chrisprice updated with d3.pairs

File: src/relativeStrengthIndex.js [hunk 3]
```diff
@@ -1,48 +1,45 @@
...
-        prevDownChangesAvg = downChangesAvg;
-        prevUpChangesAvg = upChangesAvg;
+            const closePairs = pairs([prevClose, ...closes]);
+            const downChangesAvg = downChangesAverage(closePairs.map(downChange));
+            const upChangesAvg = upChangesAverage(closePairs.map(upChange));
+            const rs = prevClose && (upChangesAvg / downChangesAvg);
 
-        const rs = upChangesAvg / downChangesAvg;
-        return convertNaN(100 - (100 / (1 + rs)));
+            prevClose = closes.reduce((prev, curr) => prev && curr, prevClose);
```
Comment (tlsim) [inline src/relativeStrengthIndex.js]: So it turns out we can assign any number here and the tests still pass. Looks like `prevClose` is only used once for the initial average.

**Comment by chrisprice** [comment]
:+1:

**Pull request merged**